srcjs/input_rate.js

```diff
@@ -211,84 +213,109 @@ var InputBatchSender = function(shinyapp) {
     const jsonValue = JSON.stringify(value);
 
-    // Resend if either json value or the input type has changed.
+    // Resend if either json value or the input type has changed. Note that if
+    // `opts` changes, it will not the value to be resent.
     if (this.lastSentValues[inputName] &&
         this.lastSentValues[inputName].jsonValue === jsonValue &&
         this.lastSentValues[inputName].inputType === inputType) {
       return;
     }
     this.lastSentValues[inputName] = { jsonValue, inputType };
-    this.target.setInput(name, value);
+    this.target.setInput(name, value, opts);
   };
   this.reset = function(values) {
     values = values || {};
-    var strValues = {};
-    $.each(values, function(key, value) {
-      strValues[key] = JSON.stringify(value);
-    });
+    var strValues = mapValues(values, x => JSON.stringify(x.value));
     this.lastSentValues = strValues;
   };
 }).call(InputNoResendDecorator.prototype);
 
+
 var InputDeferDecorator = function(target) {
   this.target = target;
   this.pendingInput = {};
 };
 (function() {
-  this.setInput = function(name, value) {
+  this.setInput = function(name, value, opts) {
+    opts = Object.assign({
+      binding: null,
+      el: null
+    }, opts);
+
     if (/^\./.test(name))
-      this.target.setInput(name, value);
+      this.target.setInput(name, value, opts);
     else
-      this.pendingInput[name] = value;
+      this.pendingInput[name] = { value, opts };
   };
   this.submit = function() {
     for (var name in this.pendingInput) {
-      if (this.pendingInput.hasOwnProperty(name))
-        this.target.setInput(name, this.pendingInput[name]);
+      if (this.pendingInput.hasOwnProperty(name)) {
+        let input = this.pendingInput[name];
+        this.target.setInput(name, input.value, input.opts);
+      }
     }
   };
 }).call(InputDeferDecorator.prototype);
 
+
 var InputEventDecorator = function(target) {
   this.target = target;
 };
 (function() {
-  this.setInput = function(name, value, immediate) {
+  this.setInput = function(name, value, opts) {
+    opts = Object.assign({
+      binding: null,
+      el: null
+    }, opts);
+
     var evt = jQuery.Event("shiny:inputchanged");
 
     const input = splitInputNameType(name);
     evt.name = input.name;
     evt.inputType = input.inputType;
     evt.value = value;
+    evt.binding   = opts.binding;
+    evt.el        = opts.el;
+
     $(document).trigger(evt);
+
     if (!evt.isDefaultPrevented()) {
       name = evt.name;
       if (evt.inputType !== '') name += ':' + evt.inputType;
-      this.target.setInput(name, evt.value, immediate);
+      this.target.setInput(name, evt.value, opts);
     }
   };
 }).call(InputEventDecorator.prototype);
 
+
 var InputRateDecorator = function(target) {
   this.target = target;
   this.inputRatePolicies = {};
 };
 (function() {
-  this.setInput = function(name, value, immediate) {
+  this.setInput = function(name, value, opts) {
```

Comment:
Instead of checking `opts` at each level of input decorator, it would be cleaner to add another level, like `inputValidatorDecorator` to the top of the stack, which checks that `name` isn't `null`, the `opts` are valid, and so on.